xenomint/state.go

```diff
@@ -0,0 +1,502 @@
+				err = ErrQueryConflict
+				return
+			}
+			continue
+		}
+		// Replay query
+		for j, v := range q.Request.Payload.Queries {
+			if q.Request.Header.QueryType == types.ReadQuery {
+				continue
+			}
+			if q.Request.Header.QueryType != types.WriteQuery {
+				err = errors.Wrapf(ErrInvalidRequest, "replay block at %d:%d", i, j)
+				s.rollbackTo(lastsp)
+				return
+			}
+			if _, ierr = s.writeSingle(&v); ierr != nil {
+				err = errors.Wrapf(ierr, "execute at %d:%d failed", i, j)
+				s.rollbackTo(lastsp)
+				return
+			}
+		}
+		s.setSavepoint()
+		s.pool.enqueue(lastsp, query)
+	}
+	// Check if the current transaction is ok to commit
+	if s.pool.matchLast(lastsp) {
+		if err = s.unc.Commit(); err != nil {
+			// FATAL ERROR
+			return
+		}
+		if s.unc, err = s.strg.Writer().Begin(); err != nil {
+			// FATAL ERROR
+			return
+		}
+		s.setNextTxID()
+	} else {
+		// Set commit point only, transaction is not actually committed. This commit point will be
+		// used on exiting.
+		s.setCommitPoint()
+	}
+	s.setSavepoint()
+	// Truncate pooled queries
+	s.pool.truncate(lastsp)
+	return
+}
+
+func (s *State) commit(out chan<- command) (err error) {
+	var queries []*QueryTracker
+	s.Lock()
+	defer s.Unlock()
+	if err = s.unc.Commit(); err != nil {
+		return
+	}
+	if s.unc, err = s.strg.Writer().Begin(); err != nil {
+		return
+	}
+	s.setNextTxID()
+	s.setSavepoint()
+	queries = s.pool.queries
+	s.pool = newPool()
+	// NOTE(leventeliu): Send commit request to the outcoming command queue within locking scope to
+	// prevent any extra writing before this commit.
+	if out != nil {
+		out <- &commitRequest{queries: queries}
+	}
+	return
+}
+
+func (s *State) CommitEx() (queries []*QueryTracker, err error) {
+	s.Lock()
+	defer s.Unlock()
+	if err = s.unc.Commit(); err != nil {
+		// FATAL ERROR
+		return
+	}
+	if s.unc, err = s.strg.Writer().Begin(); err != nil {
+		// FATAL ERROR
+		return
+	}
+	s.setNextTxID()
+	s.setSavepoint()
+	queries = s.pool.queries
+	s.pool = newPool()
+	return
+}
+
+func (s *State) partialCommit(qs []*types.Response) (err error) {
+	var (
+		i  int
+		v  *types.Response
+		q  *QueryTracker
+		rm []*QueryTracker
+
+		pl = len(s.pool.queries)
+	)
+	if len(qs) > pl {
+		err = ErrLocalBehindRemote
+		return
+	}
+	for i, v = range qs {
+		var loc = s.pool.queries[i]
+		if loc.Req.Header.Hash() != v.Header.Request.Hash() ||
+			loc.Resp.Header.LogOffset != v.Header.LogOffset {
+			err = ErrQueryConflict
+			return
+		}
+	}
+	if i < pl-1 {
+		s.rollbackTo(s.pool.queries[i+1].Resp.Header.LogOffset)
+	}
+	// Reset pool
+	rm = s.pool.queries[:i+1]
+	s.pool = newPool()
+	// Rewrite
+	for _, q = range rm {
+		if _, _, err = s.write(q.Req); err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (s *State) rollback() (err error) {
+	s.Lock()
+	defer s.Unlock()
+	s.rollbackTo(s.cmpoint)
+	s.current = s.cmpoint
+	return
+}
+
+func (s *State) Query(req *types.Request) (ref *QueryTracker, resp *types.Response, err error) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method State.Query should have comment or be unexported